File: java/advanced_ds/SuffixTree.java
package advanced_ds;

import java.util.*;

/**
 * Suffix Tree
 *
 * Time Complexity:
 * - Construction: O(n) with Ukkonen's algorithm (simplified: O(n²) here)
 * - Pattern search: O(m) where m = pattern length
 * - Longest repeated substring: O(n)
 * - Substring counting: O(n)
 * - Space Complexity: O(n)
 *
 * Use Cases:
 * - Pattern matching and substring search (faster than regex)
 * - Finding longest repeated substring
 * - Longest common substring of multiple strings
 * - String compression
 * - Palindrome detection
 *
 * Key Insight:
 * - Compact trie of all suffixes
 * - Each edge has substring, not single character
 * - Linear space and time with Ukkonen's algorithm
 * - More complex to implement than suffix array
 * - Better for multiple queries on same string
 */
public class SuffixTree {

    static class Node {
        Map<Character, Node> children = new HashMap<>();
        Node suffixLink;
        int start, end;
        int suffixIdx;

        Node() {
            this.start = -1;
            this.end = -1;
            this.suffixIdx = -1;
        }
    }

    private String text;
    private int n;
    private Node root;

    /**
     * Build suffix tree from text.
     *
     * @param text Input string
     */
    public SuffixTree(String text) {
        this.text = text + "$";  // Add sentinel
        this.n = this.text.length();
        this.root = new Node();
        this.root.start = -1;
        this.root.end = 0;

        buildTree();
    }

    /**
     * Build suffix tree by inserting all suffixes.
     */
    private void buildTree() {
        for (int i = 0; i < n; i++) {
            insertSuffix(i);
        }
    }

    /**
     * Insert suffix starting at suffixIdx.
     */
    private void insertSuffix(int suffixIdx) {
        Node node = root;
        int i = suffixIdx;

        while (i < n) {
            char c = text.charAt(i);

            if (!node.children.containsKey(c)) {
                // Create new leaf
                Node leaf = new Node();
                leaf.start = i;
                leaf.end = n;
                leaf.suffixIdx = suffixIdx;
                node.children.put(c, leaf);
                break;
            } else {
                Node edgeNode = node.children.get(c);
                int edgeLen = edgeNode.end - edgeNode.start;

                // Check if suffix matches this edge
                int matchLen = 0;
                while (matchLen < edgeLen && i + matchLen < n) {
                    if (text.charAt(edgeNode.start + matchLen) != text.charAt(i + matchLen)) {
                        break;
                    }
                    matchLen++;
                }

                if (matchLen == edgeLen) {
                    // Full edge match
                    node = edgeNode;
                    i += edgeLen;
                } else {
                    // Partial match, split edge
                    Node splitNode = new Node();
                    splitNode.start = edgeNode.start;
                    splitNode.end = edgeNode.start + matchLen;
                    splitNode.suffixIdx = -1;

                    // Old edge becomes child
                    char oldChar = text.charAt(edgeNode.start + matchLen);
                    splitNode.children.put(oldChar, edgeNode);
                    edgeNode.start = splitNode.end;

                    // New leaf
                    Node leaf = new Node();
                    leaf.start = i + matchLen;
                    leaf.end = n;
                    leaf.suffixIdx = suffixIdx;
                    splitNode.children.put(text.charAt(i + matchLen), leaf);

                    // Replace edge
                    node.children.put(c, splitNode);
                    break;
                }
            }
        }
    }

    /**
     * Find all occurrences of pattern.
     *
     * @param pattern Pattern to search
     * @return List of starting positions
     */
    public List<Integer> search(String pattern) {
        List<Integer> positions = new ArrayList<>();
        Node node = root;
        int i = 0;

        // Traverse tree following pattern
        while (i < pattern.length()) {
            char c = pattern.charAt(i);

            if (!node.children.containsKey(c)) {
                return positions;
            }

            Node edgeNode = node.children.get(c);
            String edgeStr = text.substring(edgeNode.start, edgeNode.end);
            int edgeLen = edgeStr.length();

            // Check if pattern matches this edge
            String patternRemaining = pattern.substring(i);
            int matchLen = 0;

            while (matchLen < edgeLen && matchLen < patternRemaining.length()) {
                if (edgeStr.charAt(matchLen) != patternRemaining.charAt(matchLen)) {
                    return positions;
                }
                matchLen++;
            }

            i += matchLen;
            node = edgeNode;
        }

        // Collect all leaf positions under this node
        collectPositions(node, positions);
        Collections.sort(positions);
        return positions;
    }

    /**
     * Collect suffix indices from all leaves under node.
     */
    private void collectPositions(Node node, List<Integer> positions) {
        if (node.suffixIdx != -1) {
            positions.add(node.suffixIdx);
        } else {
            for (Node child : node.children.values()) {
                collectPositions(child, positions);
            }
        }
    }

    /**
     * Find longest substring that appears at least twice.
     *
     * @return Longest repeated substring
     */
    public String longestRepeatedSubstring() {
        int[] maxLen = {0};
        Node[] maxNode = {null};

        dfs(root, 0, maxLen, maxNode);

        if (maxNode[0] != null && maxLen[0] > 0) {
            return text.substring(0, maxLen[0]);
        }

        return "";
    }

    private void dfs(Node node, int depth, int[] maxLen, Node[] maxNode) {
        if (node.suffixIdx != -1) {
            if (depth > maxLen[0]) {
                maxLen[0] = depth;
                maxNode[0] = node;
            }
        } else {
            int leafCount = countLeaves(node);
            if (leafCount >= 2 && depth > maxLen[0]) {
                maxLen[0] = depth;
                maxNode[0] = node;
            }
        }

        for (Node child : node.children.values()) {
            int childDepth = depth + (child.end - child.start);
            dfs(child, childDepth, maxLen, maxNode);
        }
    }

    /**
     * Count leaves under node.
     */
    private int countLeaves(Node node) {
        if (node.suffixIdx != -1) {
            return 1;
        }

        int count = 0;
        for (Node child : node.children.values()) {
            count += countLeaves(child);
        }

        return count;
    }

    public static void main(String[] args) {
        // Example 1: Pattern search
        System.out.println("=== Example 1: Pattern Search ===");
        String text = "banana";
        SuffixTree tree = new SuffixTree(text);

        String[] patterns = {"ana", "na", "ban", "nana"};
        System.out.println("Text: \"" + text + "\"");
        for (String pattern : patterns) {
            List<Integer> positions = tree.search(pattern);
            System.out.println("Pattern \"" + pattern + "\": positions " + positions);
        }

        // Example 2: Longest repeated substring
        System.out.println("\n=== Example 2: Longest Repeated Substring ===");
        String[] texts = {"banana", "abracadabra", "mississippi"};
        for (String t : texts) {
            SuffixTree t2 = new SuffixTree(t);
            String lrs = t2.longestRepeatedSubstring();
            System.out.println("Text \"" + t + "\": longest repeated = \"" + lrs + "\"");
        }

        // Example 3: Complex text
        System.out.println("\n=== Example 3: Complex Text ===");
        String text3 = "hello world hello";
        SuffixTree tree3 = new SuffixTree(text3);

        String[] patterns3 = {"hello", "world", "ll", "o w"};
        System.out.println("Text: \"" + text3 + "\"");
        for (String pattern : patterns3) {
            List<Integer> positions = tree3.search(pattern);
            System.out.println("Pattern \"" + pattern + "\": positions " + positions);
        }
    }
}

File: python/advanced_ds/suffix_tree.py
"""
Suffix Tree

Time Complexity:
- Construction: O(n) with Ukkonen's algorithm (simplified: O(n²) here)
- Pattern search: O(m) where m = pattern length
- Longest repeated substring: O(n)
- Substring counting: O(n)
- Space Complexity: O(n)

Use Cases:
- Pattern matching and substring search (faster than regex)
- Finding longest repeated substring
- Longest common substring of multiple strings
- String compression
- Palindrome detection

Key Insight:
- Compact trie of all suffixes
- Each edge has substring, not single character
- Linear space and time with Ukkonen's algorithm
- More complex to implement than suffix array
- Better for multiple queries on same string
"""

from typing import Optional, List, Dict, Tuple


class SuffixNode:
    """Node in suffix tree."""

    def __init__(self):
        self.children = {}
        self.suffix_link = None
        self.start = -1
        self.end = -1
        self.suffix_idx = -1


class SuffixTree:
    """Suffix tree for efficient string searching."""

    def __init__(self, text: str):
        """
        Build suffix tree from text.

        Args:
            text: Input string
        """
        self.text = text + "$"  # Add sentinel
        self.n = len(self.text)
        self.root = SuffixNode()
        self.root.start = -1
        self.root.end = 0

        # Build using simplified O(n²) approach
        self._build_tree()

    def _build_tree(self) -> None:
        """Build suffix tree by inserting all suffixes."""
        for i in range(self.n):
            self._insert_suffix(i)

    def _insert_suffix(self, suffix_idx: int) -> None:
        """Insert suffix starting at suffix_idx."""
        node = self.root
        i = suffix_idx

        while i < self.n:
            char = self.text[i]

            if char not in node.children:
                # Create new leaf
                leaf = SuffixNode()
                leaf.start = i
                leaf.end = self.n
                leaf.suffix_idx = suffix_idx
                node.children[char] = leaf
                break
            else:
                edge_node = node.children[char]
                edge_len = edge_node.end - edge_node.start

                # Check if suffix matches this edge
                match_len = 0
                while match_len < edge_len and i + match_len < self.n:
                    if self.text[edge_node.start + match_len] != self.text[i + match_len]:
                        break
                    match_len += 1

                if match_len == edge_len:
                    # Full edge match, continue with child
                    node = edge_node
                    i += edge_len
                else:
                    # Partial match, split edge
                    split_node = SuffixNode()
                    split_node.start = edge_node.start
                    split_node.end = edge_node.start + match_len
                    split_node.suffix_idx = -1

                    # Old edge becomes child
                    old_char = self.text[edge_node.start + match_len]
                    split_node.children[old_char] = edge_node
                    edge_node.start = split_node.end

                    # New leaf
                    leaf = SuffixNode()
                    leaf.start = i + match_len
                    leaf.end = self.n
                    leaf.suffix_idx = suffix_idx
                    split_node.children[self.text[i + match_len]] = leaf

                    # Replace edge
                    node.children[char] = split_node
                    break

    def search(self, pattern: str) -> List[int]:
        """
        Find all occurrences of pattern.

        Returns:
            List of starting positions
        """
        positions = []
        node = self.root
        i = 0

        # Traverse tree following pattern
        while i < len(pattern):
            char = pattern[i]

            if char not in node.children:
                return positions

            edge_node = node.children[char]
            edge_str = self.text[edge_node.start : edge_node.end]
            edge_len = len(edge_str)

            # Check if pattern matches this edge
            pattern_remaining = pattern[i:]
            match_len = 0

            while match_len < edge_len and match_len < len(pattern_remaining):
                if edge_str[match_len] != pattern_remaining[match_len]:
                    return positions
                match_len += 1

            i += match_len
            node = edge_node

        # Collect all leaf positions under this node
        self._collect_positions(node, positions)
        return sorted(positions)

    def _collect_positions(self, node: SuffixNode, positions: List[int]) -> None:
        """Collect suffix indices from all leaves under node."""
        if node.suffix_idx != -1:
            # Leaf node
            positions.append(node.suffix_idx)
        else:
            # Internal node
            for child in node.children.values():
                self._collect_positions(child, positions)

    def longest_repeated_substring(self) -> str:
        """Find longest substring that appears at least twice."""
        max_len = 0
        max_node = None

        def dfs(node: SuffixNode, depth: int = 0):

        """

        [Brief description of what this function does]


        Args:

            [param]: description


        Returns:

            [description of return value]


        Time: O([complexity])

        Space: O([complexity])

        """
            nonlocal max_len, max_node

            if node.suffix_idx != -1:
                # Leaf node
                if depth > max_len:
                    max_len = depth
                    max_node = node
            else:
                # Internal node with multiple leaves
                leaf_count = self._count_leaves(node)
                if leaf_count >= 2 and depth > max_len:
                    max_len = depth
                    max_node = node

            for child in node.children.values():
                child_depth = depth + (child.end - child.start)
                dfs(child, child_depth)

        dfs(self.root)

        if max_node and max_len > 0:
            # Reconstruct substring
            return self.text[:max_len]

        return ""

    def _count_leaves(self, node: SuffixNode) -> int:
        """Count leaves under node."""
        if node.suffix_idx != -1:
            return 1

        count = 0
        for child in node.children.values():
            count += self._count_leaves(child)

        return count


if __name__ == "__main__":
    # Example 1: Basic search
    print("=== Example 1: Pattern Search ===")
    text = "banana"
    tree = SuffixTree(text)

    patterns = ["ana", "na", "ban", "nana"]
    print(f"Text: '{text}'")
    for pattern in patterns:
        positions = tree.search(pattern)
        print(f"Pattern '{pattern}': positions {positions}")

    # Example 2: Longest repeated substring
    print("\n=== Example 2: Longest Repeated Substring ===")
    texts = ["banana", "abracadabra", "mississippi"]
    for t in texts:
        tree = SuffixTree(t)
        lrs = tree.longest_repeated_substring()
        print(f"Text '{t}': longest repeated = '{lrs}'")

    # Example 3: More complex text
    print("\n=== Example 3: Complex Text ===")
    text3 = "hello world hello"
    tree3 = SuffixTree(text3)

    patterns3 = ["hello", "world", "ll", "o w"]
    print(f"Text: '{text3}'")
    for pattern in patterns3:
        positions = tree3.search(pattern)
        print(f"Pattern '{pattern}': positions {positions}")
